### PR TITLE
fix edge case for `fn render` where resource filename starts with pkg name

### LINKED ISSUE #1691: kpt fn render doesn't handle subpackage files correctly
When we invoke `kpt fn render` on a package with subpackages in it, the rendered resources in the subpackages are written to parent parent package as opposed to writing them to subpackages itself.

e.g. for a package with following tree structure
```
wordpress
├── Kptfile
├── README.md
├── mysql
│   ├── Kptfile
│   └── mysql-deployment.yaml
└── wordpress-deployment.yaml
```

`kpt fn render` on the package results in the following

```
wordpress
├── Kptfile
├── README.md
├── mysql-deployment.yaml
├── mysql
│   ├── Kptfile
│   └── mysql-deployment.yaml
└── wordpress-deployment.yaml
```

### PR DESCRIPTION
**Pull request by droot**
This PR an edge case in `kpt fn render`. The edge case is triggered if resource file name starts with the pkgname for ex. mysql-deployment.yaml resource in a subpkg `mysql`
PR includes an e2e test case as well.

 fixes #1691

### CODE REVIEW

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  Get the example package on to local using `kpt pkg get`
+  
+  $ kpt pkg get https://github.com/GoogleContainerTools/kpt.git/package-examples/subpackages/wordpress
+  
+      fetching package /package-examples/subpackages/wordpress from https://github.com/GoogleContainerTools/kpt to wordpress
+  
+  ### View the package contents
+  List the package contents in a tree structure.
+  
+  $ kpt cfg tree wordpress/
```
Comment (phanimarupaka) [inline e2e/testdata/fn-render/resource-has-pkgname-prefix/README.md:23]: `kpt pkg tree wordpress/`

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      wordpress
+      ├── [wordpress-deployment.yaml]  Deployment wordpress
+      ├── [wordpress-deployment.yaml]  Service wordpress
+      ├── [wordpress-deployment.yaml]  PersistentVolumeClaim wp-pv-claim
+      └── Pkg: mysql
+          ├── [mysql-deployment.yaml]  PersistentVolumeClaim mysql-pv-claim
+          ├── [mysql-deployment.yaml]  Deployment wordpress-mysql
+          └── [mysql-deployment.yaml]  Service wordpress-mysql 
+  
+  ### Provide the setter values
```
Comment (phanimarupaka) [inline e2e/testdata/fn-render/resource-has-pkgname-prefix/README.md:34]: As this example is only setting annotation, you need not mention this.

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  Provide the values for all the setters.
+  
+  $ kpt fn render wordpress/
+  
+  ### Apply the package
+  
+  Apply all the contents of the package recursively to the cluster
+  
+  $ kubectl apply -f wordpress/ -R
```
Comment (phanimarupaka) [inline e2e/testdata/fn-render/resource-has-pkgname-prefix/README.md:43]: Use `kpt live apply`. Example PR https://github.com/GoogleContainerTools/kpt/pull/1692

File: internal/pipeline/executor.go [hunk 4]
```diff
@@ -281,12 +281,13 @@ func adjustRelPath(resources []*yaml.RNode, relPath string) ([]*yaml.RNode, erro
...
 	for _, r := range resources {
-		meta, err := r.GetMeta()
+		currPath, _, err := kioutil.GetFileAnnotations(r)
 		if err != nil {
 			return resources, err
 		}
-		if !strings.HasPrefix(meta.Annotations[kioutil.PathAnnotation], relPath) {
-			newPath := path.Join(relPath, meta.Annotations[kioutil.PathAnnotation])
+		// if currPath is relative to root pkg i.e. already has relPath, skip it
+		if !strings.HasPrefix(currPath, relPath+"/") {
```
Comment (phanimarupaka) [inline internal/pipeline/executor.go:289]: Shouldn't the path annotation be always relative to root package ? Why do we need to adjust that ?

**Review by phanimarupaka — CHANGES_REQUESTED** [review]

File: internal/pipeline/executor.go (hunk 4 above)
Comment (droot) [reply internal/pipeline/executor.go:289]: No, during hydration, we start from inward pkg first moving outward. So when we read (and hydrate) the resources, we read (and hydrate) them at it's pkg level (so the path is relative to its pkg) so we adjust it to the rootpkg because when we write back, we write it at the rootlevel.

So this function can get called multiple times as we move from inward towards outward. So first call to this function from the most inward pkg, will set the annotation relative to root and subsequent calls will always find the relative path there.

In our test case, `/` forward slash was missing that caused the issue.

**Review by droot — COMMENTED** [review]
I didn't intend to add `README` in this testcase, so removed it.

Thanks

**Review by phanimarupaka — APPROVED** [review]

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/.expected/diff.patch [hunk 5]
```diff
@@ -0,0 +1,199 @@
+diff --git a/mysql/mysql-deployment.yaml b/mysql/mysql-deployment.yaml
+index cbbe597..6b307e8 100644
+--- a/mysql/mysql-deployment.yaml
++++ b/mysql/mysql-deployment.yaml
+@@ -11,33 +11,36 @@
+ # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+-
```
Comment (frankfarzan) [inline e2e/testdata/fn-render/resource-has-pkgname-prefix/.expected/diff.patch:9]: can you either file a bug for this style change

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/.expected/diff.patch [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+ metadata:
+-  name: wordpress-mysql
++  annotations:
++    abc: def
++    foo: bar
++    projectId: PROJECT_ID # kpt-set: ${gcloud.core.project}
++    teamname: YOURTEAM # kpt-set: ${teamname}
+   labels:
+     app: wordpress
+-  annotations:
```
Comment (frankfarzan) [inline e2e/testdata/fn-render/resource-has-pkgname-prefix/.expected/diff.patch:21]: there's some unrelated formatting here such as reordering of the fields which obfuscates what the fuction is doing.

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile [hunk 7]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kpt.dev/v1alpha1
```
Comment (frankfarzan) [inline e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile:1]: We need a way to test packages in top-level `package-examples/` directory. They're used throughout the documentation. That also avoids duplicating it again in `e2e/testdata`.

The only reason for not having .expected in those packages is if it gets fetched the user's local filesystem and confuses them. WDYT?

File: internal/pipeline/executor.go (hunk 4 above)
Comment (frankfarzan) [inline internal/pipeline/executor.go:289]: Why use HasPrefix instead of a function in path package?

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile (hunk 7 above)
Comment (droot) [reply e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile:1]: Yes, testing `package-examples` is a MUST, we can save our users and us from troubles. You are right, we can't have .expected in those packages, it's going to be super confusing for users. One way we can solve this is:

- Have a `package-examples-expected` directory that contains the `.expected` for each example package.
- We write `TestPkgExamples` method in our e2e tests that copies the `.expected` for each test from `package-examples-expected` while staging the test and then drive the test using regular flow.

WDYT ?

File: internal/pipeline/executor.go (hunk 4 above)
Comment (droot) [reply internal/pipeline/executor.go:289]: Unfortunately, I couldn't find a method from path that can help. I think there is probably better way to solve this, so the problem we have is:
I have a relative path say relpath = `a/b` (note - without the slash) here and I have a path say currPath = `a/b/c/d` and want to find out if currPath is relative to relPath.

Comment (phanimarupaka) [reply internal/pipeline/executor.go:289]: I tried to see if there is a direct method in path to do that easily. Closest thing we can use is `filepath.Rel()` but it is not sufficient.

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile (hunk 7 above)
Comment (frankfarzan) [reply e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile:1]: I like that idea.

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/.expected/diff.patch (hunk 6 above)
Comment (phanimarupaka) [reply e2e/testdata/fn-render/resource-has-pkgname-prefix/.expected/diff.patch:21]: Seems like it is related to the functions and not `render` itself. `set-annotation`, `set-namespace`, `set-label` functions format the resources.

@Shell32-Natsu

File: e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile (hunk 7 above)
Comment (droot) [reply e2e/testdata/fn-render/resource-has-pkgname-prefix/mysql/Kptfile:1]: Filed https://github.com/GoogleContainerTools/kpt/issues/1697 to track this.
